packages/ai/core/tool/mcp/client.ts

```diff
@@ -0,0 +1,354 @@
+
+      this.responseHandlers.set(messageId, response => {
+        if (signal?.aborted) {
+          return reject(
+            new MCPClientError({
+              message: 'Request was aborted',
+              cause: signal.reason,
+            }),
+          );
+        }
+
+        if (response instanceof Error) {
+          return reject(response);
+        }
+
+        try {
+          const result = resultSchema.parse(response.result);
+          resolve(result);
+        } catch (error) {
+          reject(error);
+        }
+      });
+
+      this.transport.send(jsonrpcRequest).catch(error => {
+        cleanup();
+        reject(error);
+      });
+    });
+  }
+
+  private async listTools({
+    params,
+    options,
+  }: {
+    params?: ListToolsRequest['params'];
+    options?: RequestOptions;
+  } = {}): Promise<ListToolsResult> {
+    if (!this.serverCapabilities.tools) {
+      throw new MCPClientError({
+        message: `Server does not support tools`,
+      });
+    }
+
+    return this.request({
+      request: { method: 'tools/list', params },
+      resultSchema: ListToolsResultSchema,
+      options,
+    });
+  }
+
+  private async callTool({
+    params,
+    resultSchema,
+    options,
+  }: {
+    params: CallToolRequest['params'];
+    resultSchema: typeof CallToolResultSchema;
+    options?: RequestOptions;
+  }): Promise<CallToolResult> {
+    if (!this.serverCapabilities.tools) {
+      throw new MCPClientError({
+        message: `Server does not support tools`,
+      });
+    }
+
+    return this.request({
+      request: { method: 'tools/call', params },
+      resultSchema,
+      options,
+    });
+  }
+
+  private async notification(notification: Notification): Promise<void> {
+    const jsonrpcNotification: JSONRPCNotification = {
+      ...notification,
+      jsonrpc: '2.0',
+    };
+
+    await this.transport.send(jsonrpcNotification);
+  }
+
+  /**
+   * Returns a set of AI SDK tools from the MCP server
+   * @returns A record of tool names to their implementations
+   */
+  async tools<TOOL_SCHEMAS extends ToolSchemas = 'automatic'>({
+    schemas = 'automatic',
+  }: {
+    schemas?: TOOL_SCHEMAS;
+  } = {}): Promise<McpToolSet<TOOL_SCHEMAS>> {
+    const tools: Record<string, Tool> = {};
+    const listToolsResult = await this.listTools();
+
+    for (const { name, description, inputSchema } of listToolsResult.tools) {
+      if (schemas !== 'automatic' && !(name in schemas)) {
+        continue;
+      }
+
+      const parameters =
+        schemas === 'automatic'
+          ? jsonSchema(inputSchema as JSONSchema7)
+          : schemas[name].parameters;
+
+      const toolWithExecute = tool({
+        description,
+        parameters,
+        execute: async (
+          args: inferParameters<typeof parameters>,
+          options: ToolExecutionOptions,
+        ): Promise<CallToolResult> => {
+          options?.abortSignal?.throwIfAborted();
+
+          const result = await this.callTool({
```

Comment:
`this` can lead binding issues. it might be better to have a `self` reference in scope and use that